src/charm.py

```diff
@@ -162,6 +180,97 @@ def __init__(self, *args):
+            self.vault_autounseal_provides.on.vault_autounseal_initialize,
+            self._on_vault_autounseal_initialize,
+        )
+        self.framework.observe(
+            self.vault_autounseal_provides.on.vault_autounseal_destroy,
+            self._on_vault_autounseal_destroy,
+        )
+        self.framework.observe(
+            self.vault_autounseal_requires.on.vault_autounseal_provider_removed,
+            self._configure,
+        )
+
+    def _on_vault_autounseal_destroy(self, event: VaultAutounsealDestroy):
+        if not self.unit.is_leader():
+            return
+
+        vault = self._get_active_vault_client()
+        if vault is None:
+            logger.warning("Vault is not active, cannot disable vault autounseal")
+            return
+        vault.destroy_autounseal_credentials(event.relation.id, AUTOUNSEAL_MOUNT_PATH)
+
+    def _on_vault_autounseal_initialize(self, event: VaultAutounsealInitialize):
+        self._generate_and_set_autounseal_credentials(event.relation)
+
+    def _generate_and_set_autounseal_credentials(self, relation: Relation) -> None:
+        if not self.unit.is_leader():
+            return
+        vault = self._get_active_vault_client()
+        if vault is None:
+            logger.warning("Vault is not active, cannot generate autounseal credentials")
+            return
+
+        vault.enable_secrets_engine(SecretsBackend.TRANSIT, AUTOUNSEAL_MOUNT_PATH)
+
+        # FIXME: We should probably forward the key name in the relation data,
+        # but for now we just infer it on the requirer side, so we ignore it
+        # here.
+        key_name, approle_id, secret_id = vault.create_autounseal_credentials(
+            relation.id, AUTOUNSEAL_MOUNT_PATH
+        )
+
+        self._set_autounseal_relation_data(relation, key_name, approle_id, secret_id)
+
+    def _sync_vault_autounseal(self) -> None:
+        """Goes through all the vault-autounseal relations and sends necessary credentials."""
+        if not self.unit.is_leader():
+            logger.debug("Only leader unit can handle a vault-autounseal request")
+            return
+        outstanding_requests = self.vault_autounseal_provides.get_outstanding_requests()
+        for relation in outstanding_requests:
+            self._generate_and_set_autounseal_credentials(relation)
+
+        # FIXME: Don't disable the secrets engine until we have a plan for how
+        # to migrate auto-unseal Vaults back to manual-unseal
+        # if not self.vault_autounseal_provides.get_requirer_requests():
+        #     vault = self._get_active_vault_client()
+        #     if vault is None:
+        #         logger.warning("Vault is not active, cannot disable autounseal transit engine")
+        #         return
+        #     vault.disable_secrets_engine(AUTOUNSEAL_MOUNT_PATH)
+
+    def _set_autounseal_relation_data(
+        self, relation: Relation, key_name: str, approle_id: str, approle_secret_id: str
+    ):
```

Comment:
Can we add some docstrings? This applies at a couple of places

Reply:
To be honest I'm in a bit of a habit of not adding docstrings for private functions unless it's rather complicated.

I feel like some of them are a complete waste (see the `_set_kv_relation_data` or _get_backup_key` docstrings for example). They basically just repeats the function signature and provide nothing of value. It's just more to maintain.

But I get this isn't the place to have the discussion. I will bring it up in the retro.

Reply:
https://github.com/canonical/vault-k8s-operator/pull/365/commits/6168f1b1077f8aacf63e8cbe3bcab657d815e71e